lib/plug/static.ex

```diff
@@ -182,6 +186,100 @@ defmodule Plug.Static do
     h in only or match?({0, _}, prefix != [] and :binary.match(h, prefix))
   end
 
+  defp serve_range(conn, path, segments, [], %{gzip?: gzip?, brotli?: brotli?} = options) do
+    encoding = file_encoding(conn, path, gzip?, brotli?)
+    serve_static(encoding, segments, options)
+  end
+  defp serve_range(conn, path, segments, [range], options) do
+    encoding = file_encoding(conn, path, false, false)
+    serve_range(encoding, range, segments, options)
+  end
+  defp serve_range(_conn, _path, _segments, _range, _options) do
+    raise InvalidRangeError
+  end
+
+  @byte_range_pattern ~r/^\s*bytes=([0-9]+)?-([0-9]+)?\s*$/
+
+  defp serve_range({:ok, conn, file_info, path}, range, segments, options) do
+    file_size = elem(file_info, 1)
+
+    parsed_range =
+      @byte_range_pattern
+      |> Regex.run(range)
+      |> start_and_end(file_size)
+      |> check_bounds(file_size)
+
+    case parsed_range do
+      {range_start, range_end} ->
+        send_range(conn, path, range_start, range_end, file_size, segments, options)
+      :entire_file ->
+        serve_static({:ok, conn, file_info, path}, segments, options)
+      :invalid ->
+        raise InvalidRangeError
```

Comment:
Honestly, reading the spec, I can find support for both approaches:

[Section 3.1](https://tools.ietf.org/html/rfc7233#section-3.1):

> A server that supports range requests MAY ignore or reject a Range header field that consists of more than two overlapping ranges, or a set of many small ranges that are not listed in ascending order, since both are indications of either a broken client or a deliberate denial-of-service attack (Section 6.1).

*or* (later in same section):

> If all of the preconditions are true, the server supports the Range header field for the target resource, and the specified range(s) are invalid or unsatisfiable, the server SHOULD send a 416 (Range Not Satisfiable) response.

[Section 4.4](https://tools.ietf.org/html/rfc7233#section-4.4):

> The 416 (Range Not Satisfiable) status code indicates that none of the ranges in the request's Range header field (Section 3.1) overlap the current extent of the selected resource or that the set of ranges requested has been rejected due to invalid ranges or an excessive request of small or overlapping ranges.
>
> For byte ranges, failing to overlap the current extent means that the first-byte-pos of all of the byte-range-spec values were greater than the current length of the selected representation.  When this status code is generated in response to a byte-range request, the sender SHOULD generate a Content-Range header field specifying the current length of the selected representation (Section 4.2).

I think the argument being made here is that returning the entire file may be "friendlier" than returning a 416, and I could probably go along with that. I'll put it in a separate commit within this PR so we can revert it if you disagree.